### PR TITLE
core/ajax/preview.php to controller

### PR DESCRIPTION
**Pull request by rullzer**
Yet another ajax endpoint that can go. And thus functionality we can actually test!

At the moment I dropped the forceIcon argument. Since it makes the code ugly. And we should not serve tiny png files anyway. Does anybody know if the forceIcon is actually used? (the files app already sets it to 0).

Unit tests are added and coverage is at 100%.

CC: @PVince81 @nickvergessen @Xenopathic @MorrisJobke

### CODE REVIEW

**Comment by rullzer** [comment]
@owncloud-bot retest this please

**Comment by karlitschek** [comment]
@rullzer https://github.com/owncloud/core/search?q=forceIcon&type=Issues&utf8=✓

**Comment by oparoz** [comment]
Great initiative :)

Maybe move this to the app domain (of core) instead of keeping it into core since making Preview an app is on @georgehrke's roadmap for 8.2?

**Comment by georgehrke** [comment]
> Does anybody know if the forceIcon is actually used?

forceIcon defaults to true, so we have to check every place a preview url is called ...
Trashbin or public sharing might use it, not sure though.
I'll check tomorrow (if no one is faster than me)

> Maybe move this to the app domain (of core) instead of keeping it into core since making Preview an app is on @georgehrke's roadmap for 8.2?

No, was outsourcing previews to an app discussed somewhere or is there an issue for that?

**Comment by oparoz** [comment]
@georgehrke https://github.com/owncloud/core/issues/13610
And there were discussions before that which led to the issue creation and which were referring to the use of an app for everything related to previews.

**Comment by georgehrke** [comment]
In this issue I meant moving it to a controller and getting rid of the ajax end points (just as this pr does).
I never wanted to move the preview functionality to an app. Sorry if there was some confusion about it.

**Comment by oparoz** [comment]
OK, but why not keep all appframework based stuff in the apps folder? /core contains old ajax endpoints, assets, templates. It doesn't seem right to add Preview there.

**Comment by rullzer** [comment]
avatar stuff is also in /core

**Comment by oparoz** [comment]
> avatar stuff is also in /core

@rullzer Yeah, one controller. That's why I'm thinking it wouldn't be a bad thing to start the trend, but the project's architects have to voice their opinion.

**Comment by georgehrke** [comment]
Because the preview api is an api offered by ownCloud's core. If we'd move it to an app, we would have apps depending on other apps. That's something we try to avoid.
The gallery app may not a dependency on some files_preview app or what soever.

**Comment by oparoz** [comment]
I'm still not convinced because the provisioning API is an app and the public sharing API is also in an app. So we end up with APIs being scattered around.

**Comment by PVince81** [comment]
@oparoz good point. This also reminds me that previews can be disabled in the config.
So if "previews" were simply an additional plugin for ownCloud, disabling that app would disable previews everywhere.

But it also means that the previews app knows how to inject previews in various apps like files_versions, sidebar, etc. Or alternatively have other apps check if there is a preview provider and get a preview from there. This sounds like a bigger change that needs to be thought through.

**Comment by rullzer** [comment]
@PVince81 with previews that would be "simple" since the dependency is via GET request. so if it is not there it will return a 404... just like the PR does when there is no preview :P

**Comment by oparoz** [comment]
@PVince81 - I agree that this might require a bigger discussion, but iirc, the end goal is to make previews completely optional, for @LukasReschke reasons, and there is a fallback mechanism with the media type icons, so user can always kind of see of which type each file is. They just don't get the previews.
So, Preview could perfectly exist as an official app, just like Documents.
The problem I see with ownCloud's strategy is that providers are supposed to also be moved outside of core, to the app store, effectively creating dependencies between lots of apps.

**Comment by georgehrke** [comment]
Well, but moving it to a dedicated app would create some really messed up dependencies.
Preview would depend on files, files_versions, files_trashbin, files_sharing, ... because it would have to inject code to display previews.
On the other hand gallery would have a dependency on previews.

> providers are supposed to also be moved outside of core

Some providers, but not all. We want to outsource some of them / make the preview system extendable, because it would be insane to have preview providers for everything. (like some file formats used in hospitals or whatsoever)

**Comment by oparoz** [comment]
Maybe we're talking about 2 different things?

lib/private/preview.php stays where it is. That's what apps are using when they're injecting the dependencies. Gallery currently does that.

The Preview app/API would use the same class, but provide REST endpoints for apps which need it. Like Gallery at some point. That's the app created in this PR and it can sit anywhere, not necessarily in core. Other apps connect and either get a preview or a media type icon.

Or am I still missing something?

File: core/preview/previewcontroller.php [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+	/**
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 *
+	 * @param string $file
+	 * @param int $x The max X
+	 * @param int $y The max Y
+	 * @param bool $scalingUp
+	 * @param bool $a Keep aspect ratio
+	 * @param bool $forceIcon
```
Comment (nickvergessen) [inline core/preview/previewcontroller.php]: not part of the method

**Comment by DeepDiver1975** [comment]
Previews will be part of the next Generation webdav endpoint in 9.0

Kind of mit neccesary to move to a controller now.

File: core/application.php [hunk 2]
```diff
@@ -88,6 +89,14 @@ public function __construct(array $urlParams=array()){
 				$c->query('UserFolder')
 			);
 		});
+		$container->registerService('PreviewController', function(SimpleContainer $c) {
```
Comment (georgehrke) [inline core/application.php]: How reusable is this?
Creating the endpoint in the trash bin app should just be a matter of replacing the folder that's being handed over.

**Comment by oparoz** [comment]
> Kind of mit neccesary to move to a controller now.

OK :)

**Comment by rullzer** [comment]
> Kind of mit neccesary to move to a controller now.

Ok lets close this then and keep previews the way it is.

**Comment by oparoz** [comment]
This PR brings needed features like proper return codes and aspect ratio support. On top of that it's fully tested.
Would it be that bad to bring this in for 8.2?
Without it per example, failed previews are fugly in a lot of places, like activity or the sidebar because they're stretched 16x16 PNGs.

**Pull request closed without merging**